src/Storages/StorageS3Settings.cpp

```diff
@@ -12,60 +13,232 @@
+
+    if (!if_changed || settings.s3_upload_part_size_multiply_parts_count_threshold.changed)
+        upload_part_size_multiply_parts_count_threshold = settings.s3_upload_part_size_multiply_parts_count_threshold;
+
+    if (!if_changed || settings.s3_max_single_part_upload_size.changed)
+        max_single_part_upload_size = settings.s3_max_single_part_upload_size;
+}
 
-    auto get_uint_for_key = [&](const String & key, const String & elem, bool with_default = true, UInt64 default_value = 0)
+void S3Settings::RequestSettings::PartUploadSettings::validate()
+{
+    static constexpr size_t min_upload_part_size_limit = 5 * 1024 * 1024;
+    if (min_upload_part_size < min_upload_part_size_limit)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting min_upload_part_size has invalid value {} which is less than the s3 API limit {}",
+            ReadableSize(min_upload_part_size), ReadableSize(min_upload_part_size_limit));
+
+    static constexpr size_t max_upload_part_size_limit = 5ull * 1024 * 1024 * 1024;
+    if (max_upload_part_size > max_upload_part_size_limit)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting max_upload_part_size has invalid value {} which is grater than the s3 API limit {}",
+            ReadableSize(max_upload_part_size), ReadableSize(max_upload_part_size_limit));
+
+    if (max_single_part_upload_size > max_upload_part_size_limit)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting max_single_part_upload_size has invalid value {} which is grater than the s3 API limit {}",
+            ReadableSize(max_single_part_upload_size), ReadableSize(max_upload_part_size_limit));
+
+    if (max_single_operation_copy_size > max_upload_part_size_limit)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting max_single_operation_copy_size has invalid value {} which is grater than the s3 API limit {}",
+            ReadableSize(max_single_operation_copy_size), ReadableSize(max_upload_part_size_limit));
+
+    if (max_upload_part_size < min_upload_part_size)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting max_upload_part_size ({}) can't be less than setting min_upload_part_size {}",
+            ReadableSize(max_upload_part_size), ReadableSize(min_upload_part_size));
+
+    if (!upload_part_size_multiply_factor)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting upload_part_size_multiply_factor cannot be zero",
+            upload_part_size_multiply_factor);
+
+    if (!upload_part_size_multiply_parts_count_threshold)
+        throw Exception(
+            ErrorCodes::INVALID_SETTING_VALUE,
+            "Setting upload_part_size_multiply_parts_count_threshold cannot be zero",
+            upload_part_size_multiply_parts_count_threshold);
+
+    if (!max_part_number)
```

Comment:
(Not directly related to this PR but it would be nice if it was possible to specify min/max values for each setting in Core/Settings.h. Invalid settings could then be rejected right away and not only during a `validate()` call.)